handlers.go

```diff
@@ -311,21 +311,27 @@ func (s Server) schemasHandler(w http.ResponseWriter, r *http.Request) {
 		errorHandler(w, r, paramsErr)
 		return
 	}
-	filter := f.NewFilter(params.Filter, schema.Definition())
 
-	schemas := s.getSchemas()
-	start, end := clamp(params.StartIndex-1, params.Count, len(schemas))
-	var resources []interface{}
+	var (
+		validator  = f.NewFilterValidator(params.Filter, schema.Definition())
+		schemas    = s.getSchemas()
+		start, end = clamp(params.StartIndex-1, params.Count, len(schemas))
+		resources  []interface{}
+	)
+
+	if params.Filter != nil {
+		if err := validator.Validate(); err != nil {
+			errorHandler(w, r, &errors.ScimErrorInvalidFilter)
+			return
+		}
+	}
 	for _, v := range schemas[start:end] {
 		resource := v.ToMap()
 		if params.Filter != nil {
-			valid, err := filter.IsValid(resource)
-			if err != nil {
-				scimErr := errors.CheckScimError(err, http.MethodGet)
-				errorHandler(w, r, &scimErr)
-				return
-			}
-			if !valid {
+			if err := validator.PassesFilter(resource); err != nil {
+				if err, ok := err.(*errors.ScimError); ok {
+					panic(err)
+				}
```

Comment:
This is becoming really hard to understand IMO. Why not `panic` in the validator itself, immediately when something unexpected happens?